File: backend/src/modules/user/user.service.ts
import { Injectable } from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import { PrismaService } from '../../prisma/prisma.service';
import { decryptToken, encryptToken } from '../../common/crypto/token.crypto';
import { fail, success, successList, successWrite, paginationHeader } from '../../common/response/base-response';
import { LoginDto } from './dto/login.dto';

/** Payload for creating / editing a login user from the Users admin screen. */
export interface UserAdminDto {
  Id?: number;
  Name?: string;
  Email?: string;
  Mobile?: string;
  Password?: string;
  RoleId?: number;
  Active?: boolean;
}

@Injectable()
export class UserService {
  constructor(private readonly prisma: PrismaService) {}

  private companyAllowed(name?: string): boolean {
    const allowed = (process.env.ALLOWED_COMPANIES ?? 'demo')
      .split(',')
      .map((s) => s.trim().toLowerCase());
    return !!name && allowed.includes(name.toLowerCase());
  }

  private fullName(u: { firstName?: string | null; middleName?: string | null; lastName?: string | null }): string {
    return [u.firstName, u.middleName, u.lastName].filter(Boolean).join(' ');
  }

  /** GET /User/GetAllUsers — active LOGIN users (User table) with their roles,
   *  for pickers such as the route-supervisor dropdown. These are the accounts
   *  that log in and hold roles — distinct from HrUser/passenger records. */
  async getAllUsers() {
    const users = await this.prisma.user.findMany({
      where: { active: true },
      include: { userRoles: { where: { active: true }, include: { role: true } } },
      orderBy: { id: 'asc' },
    });
    return success(
      users.map((u) => ({
        Id: u.id,
        Name: this.fullName(u) || u.email,
        RoleList: u.userRoles.map((r) => ({ RoleID: r.roleId, RoleName: r.role.name })),
      })),
    );
  }

  /** Split a display name into first / last for the User table (single-field UI). */
  private splitName(name?: string): { firstName: string; lastName: string | null } {
    const parts = (name ?? '').trim().split(/\s+/).filter(Boolean);
    if (parts.length === 0) return { firstName: '', lastName: null };
    if (parts.length === 1) return { firstName: parts[0], lastName: null };
    return { firstName: parts[0], lastName: parts.slice(1).join(' ') };
  }

  /** GET /User/GetUserList — paginated login users (optionally filtered by role)
   *  for the Users administration screen. Each row carries its role list. */
  async getUserList(pageNo = 1, noOfItems = 20, roleId?: number) {
    const where: Record<string, unknown> = { active: true };
    if (roleId) where.userRoles = { some: { roleId, active: true } };

    const [total, users] = await this.prisma.$transaction([
      this.prisma.user.count({ where }),
      this.prisma.user.findMany({
        where,
        include: { userRoles: { where: { active: true }, include: { role: true } } },
        orderBy: { id: 'asc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);

    const data = users.map((u) => ({
      Id: u.id,
      Name: this.fullName(u) || u.email,
      Email: u.email,
      Mobile: u.mobile ?? '',
      Active: u.active,
      RoleList: u.userRoles.map((r) => ({ RoleID: r.roleId, RoleName: r.role.name })),
    }));
    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** POST /User/AddUser — create a login user and assign a transportation role. */
  async addUser(dto: UserAdminDto) {
    if (!dto?.Email || !dto?.Password) return fail('Err101', 'Email and Password are required');
    if (!dto?.RoleId) return fail('Err101', 'RoleId is required');
    const existing = await this.prisma.user.findFirst({ where: { email: dto.Email } });
    if (existing) return fail('Err102', 'A user with this email already exists');

    const { firstName, lastName } = this.splitName(dto.Name);
    const password = await bcrypt.hash(dto.Password, 10);
    const user = await this.prisma.user.create({
      data: {
        email: dto.Email,
        password,
        firstName,
        lastName,
        mobile: dto.Mobile ?? null,
        active: dto.Active ?? true,
        userRoles: { create: [{ roleId: dto.RoleId, active: true }] },
      },
    });
    return successWrite(user.id, 'User created');
  }

  /** POST /User/UpdateUser — edit a login user's profile, role and status. */
  async updateUser(dto: UserAdminDto) {
    if (!dto?.Id) return fail('Err101', 'Id is required');
    const user = await this.prisma.user.findUnique({ where: { id: dto.Id } });
    if (!user) return fail('Err142', 'User not found');

    // Guard against stealing another account's email.
    if (dto.Email && dto.Email !== user.email) {
      const clash = await this.prisma.user.findFirst({ where: { email: dto.Email, id: { not: dto.Id } } });
      if (clash) return fail('Err102', 'A user with this email already exists');
    }

    const data: Record<string, unknown> = {};
    if (dto.Name !== undefined) {
      const { firstName, lastName } = this.splitName(dto.Name);
      data.firstName = firstName;
      data.middleName = null;
      data.lastName = lastName;
    }
    if (dto.Email !== undefined) data.email = dto.Email;
    if (dto.Mobile !== undefined) data.mobile = dto.Mobile || null;
    if (dto.Active !== undefined) data.active = dto.Active;
    if (dto.Password) data.password = await bcrypt.hash(dto.Password, 10);
    await this.prisma.user.update({ where: { id: dto.Id }, data });

    // Reconcile the role: deactivate any others, (re)activate the chosen one.
    if (dto.RoleId) {
      await this.prisma.userRole.updateMany({
        where: { userId: dto.Id, roleId: { not: dto.RoleId }, active: true },
        data: { active: false },
      });
      const current = await this.prisma.userRole.findFirst({ where: { userId: dto.Id, roleId: dto.RoleId } });
      if (current) {
        if (!current.active) await this.prisma.userRole.update({ where: { id: current.id }, data: { active: true } });
      } else {
        await this.prisma.userRole.create({ data: { userId: dto.Id, roleId: dto.RoleId, active: true } });
      }
    }
    return successWrite(dto.Id, 'User updated');
  }

  /** POST /User/DeleteUser — deactivate a login user and end its sessions. */
  async deleteUser(id: number) {
    if (!id) return fail('Err101', 'Id is required');
    const user = await this.prisma.user.findUnique({ where: { id } });
    if (!user) return fail('Err142', 'User not found');
    await this.prisma.user.update({ where: { id }, data: { active: false } });
    await this.prisma.userSession.updateMany({ where: { userId: id, active: true }, data: { active: false } });
    return successWrite(id, 'User deactivated');
  }

  /** POST /User/Login — validates company + credentials, opens a session,
   *  returns the documented LoginResponse (Data = UserToken, RoleList, …). */
  async login(dto: LoginDto) {
    if (!dto?.Email || (!dto?.Password && dto?.ExternalLoginFrom !== 'office365')) {
      return fail('Err101', 'Email and Password are required');
    }
    if (!this.companyAllowed(dto.CompanyName)) {
      return fail('Err-P200', 'Invalid or unknown CompanyName');
    }

    const user = await this.prisma.user.findFirst({
      where: { email: dto.Email, active: true },
      include: { hrUser: true },
    });
    if (!user) return fail('Err-P6', 'Invalid credentials');

    // ⚠ office365 path skips the password check — faithful to the old docs (insecure).
    if (dto.ExternalLoginFrom !== 'office365') {
      const okPass = await bcrypt.compare(dto.Password ?? '', user.password);
      if (!okPass) return fail('Err-P6', 'Invalid credentials');
    }

    const hours = Number(process.env.SESSION_HOURS ?? 24);
    const session = await this.prisma.userSession.create({
      data: { userId: user.id, active: true, endDate: new Date(Date.now() + hours * 3600_000) },
    });
    const token = encodeURIComponent(encryptToken(String(session.id)));

    const roles = await this.prisma.userRole.findMany({
      where: { userId: user.id, active: true },
      include: { role: true },
    });

    return {
      Result: true,
      Errors: [],
      Data: token, // LoginResponse.Data = UserToken
      UserID: encryptToken(String(user.id)),
      UserIDNO: user.id,
      Name: this.fullName(user),
      PhotoUrl: user.photoUrl ?? '',
      BranchId: user.branchId ?? null,
      EmplyeeId: user.hrUser?.id ?? null, // ⚠ frozen misspelling: linked HrUser id
      RoleList: roles.map((r) => ({ RoleID: r.roleId, RoleName: r.role.name })),
      GroupList: [],
    };
  }

  /** POST /User/Logout — deactivates the session behind the token. */
  async logout(token?: string) {
    if (!token) return fail('Err142', 'Token required');
    try {
      const sessionId = Number(decryptToken(decodeURIComponent(token)));
      await this.prisma.userSession.updateMany({ where: { id: sessionId }, data: { active: false } });
    } catch {
      /* malformed token — treat as already logged out */
    }
    return { Result: true, Errors: [] };
  }

  /** GET /User/GetUserData — rebuild identity for an active session. */
  async getUserData(userId: number) {
    const user = await this.prisma.user.findUnique({ where: { id: userId }, include: { hrUser: true } });
    if (!user) return fail('Err142', 'User not found');
    const roles = await this.prisma.userRole.findMany({
      where: { userId, active: true },
      include: { role: true },
    });
    return success({
      UserIDNO: user.id,
      Name: this.fullName(user),
      PhotoUrl: user.photoUrl ?? '',
      BranchId: user.branchId ?? null,
      EmplyeeId: user.hrUser?.id ?? null,
      RoleList: roles.map((r) => ({ RoleID: r.roleId, RoleName: r.role.name })),
    });
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/User/UserService.cs
using GarasTransport.Api.Common;
using Microsoft.EntityFrameworkCore;
using BCryptNet = BCrypt.Net.BCrypt;
using UserEntity = GarasTransport.Domain.Entities.User;

namespace GarasTransport.Api.Modules.User;

public class UserService
{
    private readonly AppDbContext _db;
    private readonly TokenCrypto _crypto;
    private readonly IConfiguration _config;
    private readonly GarasTransport.Infrastructure.Tenancy.TenantRegistry _tenants;

    public UserService(AppDbContext db, TokenCrypto crypto, IConfiguration config, GarasTransport.Infrastructure.Tenancy.TenantRegistry tenants)
    {
        _db = db;
        _crypto = crypto;
        _config = config;
        _tenants = tenants;
    }

    private static string FullName(UserEntity u) =>
        string.Join(' ', new[] { u.FirstName, u.MiddleName, u.LastName }.Where(s => !string.IsNullOrEmpty(s)));

    /// <summary>Split a display name into first / last for the User table (single-field UI).</summary>
    private static (string FirstName, string? LastName) SplitName(string? name)
    {
        var parts = (name ?? string.Empty).Trim()
            .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        if (parts.Length == 0) return (string.Empty, null);
        if (parts.Length == 1) return (parts[0], null);
        return (parts[0], string.Join(' ', parts.Skip(1)));
    }

    /// <summary>GET /User/GetAllUsers — active LOGIN users with their roles, for pickers.</summary>
    public async Task<object> GetAllUsers()
    {
        var users = await _db.Users
            .Where(u => u.Active)
            .Include(u => u.UserRoles.Where(r => r.Active)).ThenInclude(r => r.Role)
            .OrderBy(u => u.Id)
            .ToListAsync();

        return Envelope.Success(users.Select(u => new
        {
            Id = u.Id,
            Name = string.IsNullOrEmpty(FullName(u)) ? u.Email : FullName(u),
            RoleList = u.UserRoles.Select(r => new { RoleID = r.RoleId, RoleName = r.Role!.Name }),
        }));
    }

    /// <summary>GET /User/GetUserList — paginated login users (optionally by role).</summary>
    public async Task<object> GetUserList(int pageNo = 1, int noOfItems = 20, int? roleId = null)
    {
        var query = _db.Users.Where(u => u.Active);
        if (roleId.HasValue)
            query = query.Where(u => u.UserRoles.Any(r => r.RoleId == roleId.Value && r.Active));

        var total = await query.CountAsync();
        var users = await query
            .Include(u => u.UserRoles.Where(r => r.Active)).ThenInclude(r => r.Role)
            .OrderBy(u => u.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .ToListAsync();

        var data = users.Select(u => new
        {
            Id = u.Id,
            Name = string.IsNullOrEmpty(FullName(u)) ? u.Email : FullName(u),
            Email = u.Email,
            Mobile = u.Mobile ?? string.Empty,
            Active = u.Active,
            RoleList = u.UserRoles.Select(r => new { RoleID = r.RoleId, RoleName = r.Role!.Name }),
        });
        return Envelope.SuccessList(data, Envelope.Pagination(pageNo, noOfItems, total));
    }

    // Requested roles: RoleIds list wins; single RoleId is the legacy fallback.
    private static List<int> RequestedRoles(UserAdminDto dto) =>
        (dto.RoleIds is { Count: > 0 } ? dto.RoleIds : dto.RoleId.HasValue ? new List<int> { dto.RoleId.Value } : new List<int>())
            .Distinct().ToList();

    /// <summary>POST /User/AddUser — create a login user and assign one or more roles.</summary>
    public async Task<object> AddUser(UserAdminDto dto)
    {
        if (string.IsNullOrEmpty(dto.Email) || string.IsNullOrEmpty(dto.Password))
            return Envelope.Fail("Err101", "Email and Password are required");
        var roleIds = RequestedRoles(dto);
        if (roleIds.Count == 0)
            return Envelope.Fail("Err101", "RoleId is required");
        if (await _db.Users.AnyAsync(u => u.Email == dto.Email))
            return Envelope.Fail("Err102", "A user with this email already exists");

        var (firstName, lastName) = SplitName(dto.Name);
        var user = new UserEntity
        {
            Email = dto.Email!,
            Password = BCryptNet.HashPassword(dto.Password, 10),
            FirstName = firstName,
            LastName = lastName,
            Mobile = dto.Mobile,
            Active = dto.Active ?? true,
            UserRoles = roleIds.Select(id => new GarasTransport.Domain.Entities.UserRole { RoleId = id, Active = true }).ToList(),
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(user.Id, "User created");
    }

    /// <summary>POST /User/UpdateUser — edit a login user's profile, role and status.</summary>
    public async Task<object> UpdateUser(UserAdminDto dto)
    {
        if (!dto.Id.HasValue) return Envelope.Fail("Err101", "Id is required");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == dto.Id.Value);
        if (user == null) return Envelope.Fail("Err142", "User not found");

        // Guard against stealing another account's email.
        if (!string.IsNullOrEmpty(dto.Email) && dto.Email != user.Email)
        {
            var clash = await _db.Users.AnyAsync(u => u.Email == dto.Email && u.Id != dto.Id.Value);
            if (clash) return Envelope.Fail("Err102", "A user with this email already exists");
        }

        if (dto.Name != null)
        {
            var (firstName, lastName) = SplitName(dto.Name);
            user.FirstName = firstName;
            user.MiddleName = null;
            user.LastName = lastName;
        }
        if (dto.Email != null) user.Email = dto.Email;
        if (dto.Mobile != null) user.Mobile = string.IsNullOrEmpty(dto.Mobile) ? null : dto.Mobile;
        if (dto.Active.HasValue) user.Active = dto.Active.Value;
        if (!string.IsNullOrEmpty(dto.Password)) user.Password = BCryptNet.HashPassword(dto.Password, 10);
        await _db.SaveChangesAsync();

        // Reconcile roles against the requested set: deactivate the ones no
        // longer wanted, (re)activate/add every requested one. Omitting both
        // RoleIds and RoleId leaves the roles untouched.
        var roleIds = RequestedRoles(dto);
        if (roleIds.Count > 0)
        {
            var existing = await _db.UserRoles
                .Where(r => r.UserId == dto.Id.Value)
                .ToListAsync();
            foreach (var o in existing.Where(r => r.Active && !roleIds.Contains(r.RoleId)))
                o.Active = false;
            foreach (var id in roleIds)
            {
                var current = existing.FirstOrDefault(r => r.RoleId == id);
                if (current != null) current.Active = true;
                else _db.UserRoles.Add(new GarasTransport.Domain.Entities.UserRole { UserId = dto.Id.Value, RoleId = id, Active = true });
            }
            await _db.SaveChangesAsync();
        }
        return Envelope.SuccessWrite(dto.Id.Value, "User updated");
    }

    /// <summary>POST /User/DeleteUser — deactivate a login user and end its sessions.</summary>
    public async Task<object> DeleteUser(int id)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == id);
        if (user == null) return Envelope.Fail("Err142", "User not found");
        user.Active = false;
        var sessions = await _db.UserSessions.Where(s => s.UserId == id && s.Active).ToListAsync();
        foreach (var s in sessions) s.Active = false;
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id, "User deactivated");
    }

    /// <summary>POST /User/Login — validates company + credentials, opens a session.</summary>
    public async Task<object> Login(LoginDto dto)
    {
        if (string.IsNullOrEmpty(dto.Email) || (string.IsNullOrEmpty(dto.Password) && dto.ExternalLoginFrom != "office365"))
            return Envelope.Fail("Err101", "Email and Password are required");
        // Login carries CompanyName in the BODY (no header yet), so the injected
        // DbContext started on the default connection. Re-target it to the
        // company's own database BEFORE the first query — database-per-company.
        if (!_tenants.TryGetConnectionString(dto.CompanyName, out var tenantCs))
            return Envelope.Fail("Err-P200", "Invalid or unknown CompanyName");
        _db.Database.SetConnectionString(tenantCs);

        var user = await _db.Users
            .Include(u => u.HrUser)
            .FirstOrDefaultAsync(u => u.Email == dto.Email && u.Active);
        if (user == null) return Envelope.Fail("Err-P6", "Invalid credentials");

        // ⚠ office365 path skips the password check — faithful to the old docs (insecure).
        if (dto.ExternalLoginFrom != "office365")
        {
            var okPass = BCryptNet.Verify(dto.Password ?? string.Empty, user.Password);
            if (!okPass) return Envelope.Fail("Err-P6", "Invalid credentials");
        }

        var hours = int.TryParse(_config["SessionHours"], out var h) ? h : 24;
        var session = new GarasTransport.Domain.Entities.UserSession
        {
            UserId = user.Id,
            Active = true,
            EndDate = DateTime.UtcNow.AddHours(hours),
        };
        _db.UserSessions.Add(session);
        await _db.SaveChangesAsync();
        // Token carries "sessionId|company" so a token can never authenticate
        // against another company's database (ids repeat across tenant DBs).
        var token = Uri.EscapeDataString(_crypto.EncryptSession(session.Id, dto.CompanyName));

        var roles = await _db.UserRoles
            .Include(r => r.Role)
            .Where(r => r.UserId == user.Id && r.Active)
            .ToListAsync();

        return new
        {
            Result = true,
            Errors = Array.Empty<ErrorItem>(),
            Data = token, // LoginResponse.Data = UserToken
            UserID = _crypto.Encrypt(user.Id.ToString()),
            UserIDNO = user.Id,
            Name = FullName(user),
            PhotoUrl = user.PhotoUrl ?? string.Empty,
            BranchId = (int?)user.BranchId,
            EmplyeeId = (int?)user.HrUser?.Id, // ⚠ frozen misspelling: linked HrUser id
            RoleList = roles.Select(r => new { RoleID = r.RoleId, RoleName = r.Role!.Name }),
            GroupList = Array.Empty<object>(),
        };
    }

    /// <summary>POST /User/Logout — deactivates the session behind the token.</summary>
    public async Task<object> Logout(string? token)
    {
        if (string.IsNullOrEmpty(token)) return Envelope.Fail("Err142", "Token required");
        try
        {
            if (_crypto.TryDecryptSession(Uri.UnescapeDataString(token), out var sessionId, out var company) && sessionId != 0)
            {
                // End the session in the company's own database (token is tenant-bound).
                if (_tenants.TryGetConnectionString(company, out var cs))
                    _db.Database.SetConnectionString(cs);
                var sessions = await _db.UserSessions.Where(s => s.Id == sessionId).ToListAsync();
                foreach (var s in sessions) s.Active = false;
                await _db.SaveChangesAsync();
            }
        }
        catch
        {
            /* malformed token — treat as already logged out */
        }
        return new BaseResponse { Result = true, Errors = new() };
    }

    /// <summary>GET /User/GetUserData — rebuild identity for an active session.</summary>
    public async Task<object> GetUserData(int userId)
    {
        var user = await _db.Users.Include(u => u.HrUser).FirstOrDefaultAsync(u => u.Id == userId);
        if (user == null) return Envelope.Fail("Err142", "User not found");
        var roles = await _db.UserRoles
            .Include(r => r.Role)
            .Where(r => r.UserId == userId && r.Active)
            .ToListAsync();
        return Envelope.Success(new
        {
            UserIDNO = user.Id,
            Name = FullName(user),
            PhotoUrl = user.PhotoUrl ?? string.Empty,
            BranchId = (int?)user.BranchId,
            EmplyeeId = (int?)user.HrUser?.Id,
            RoleList = roles.Select(r => new { RoleID = r.RoleId, RoleName = r.Role!.Name }),
        });
    }
}
